Add support for GoogleAPI OAuth service

martindufort opened the pull request:
Add support for the GoogleAPI Oauth service.

martindufort committed:
Add support for GoogleAPI OAuth service

martindufort committed:
Merge remote-tracking branch 'upstream/main' into feature/GoogleAPI

martindufort committed:
Update documentation (Readme) to include reference to Google API.

martindufort committed:
Update documentation (Readme) to include reference to Google API.

martindufort committed:
Change visibility of accessor for Login information within the Authenticator

martindufort committed:
Merge branch 'feature/PublicLogin' into feature/GoogleAPI

mattmassicotte commented on Sources/OAuthenticator/Authenticator.swift:
```diff
@@ -147,15 +149,15 @@ public final class Authenticator {
 
 extension Authenticator {
-	private func retrieveLogin() async throws -> Login? {
+	public func retrieveLogin() async throws -> Login? {
```
I'm open to this change, but I still would like to discuss it separately to make sure I fully understand why it is needed. Are you ok with omitting it from this PR?

mattmassicotte commented on Sources/OAuthenticator/Authenticator.swift:
```diff
@@ -147,15 +149,15 @@ public final class Authenticator {
 	}
 
-	private func storeLogin(_ login: Login) async throws {
+	public func storeLogin(_ login: Login) async throws {
```
Same with this.

mattmassicotte commented on README.md:
```diff
@@ -134,6 +135,62 @@ let request = URLRequest(url: url)
+
+// 2: Access secure Google endpoint (ie: Google Drive: upload a file) with access token
+var urlBuilder = URLComponents()
+urlBuilder.scheme = GoogleDrive.scheme            // https:
```
Shouldn't this be `GoogleAPI`?

mattmassicotte commented on Sources/OAuthenticator/Services/GoogleAPI.swift:
```diff
@@ -0,0 +1,221 @@
+            // For now, just display that less scope was authorized
+            os_log(.info, "[Authentication] Granted scopes less than requested scopes")
+            print("GrantedScope: \(grantedScope)")
+            print("RequestedScope: \(appCredentials.scopeString)")
```
Did you intend on leaving this logging here?

mattmassicotte commented on Sources/OAuthenticator/Services/GoogleAPI.swift:
```diff
@@ -0,0 +1,221 @@
+
+public extension URL {
+    /// Add the ability to retrieve the scope query item from the URL
+    var grantedScope: String {
```
This is an interesting property! There are a few others in URL+Code.swift. Should we maybe rename that file and move more in there?

martindufort replied:
Good idea. I have move this code into a renamed `URL+QueryParams` file.

mattmassicotte commented on Tests/OAuthenticatorTests/GoogleTests.swift:
```diff
@@ -0,0 +1,39 @@
+
+import XCTest
+import OSLog
+@testable import OAuthenticator
```
Is the `@testable` needed here?

martindufort replied:
If I don't specify `@testable` than line 24 generates an error

mattmassicotte commented:
This is amazing, and I'm very grateful that you put this together. Just a few minor things I brought up.

martindufort commented:
Thanks for the review @mattmassicotte. 
I will submit another PR that addresses your comments early next week. 

As for the `retrieveLogin()` public accessor, I'm glad to discuss and perhaps my use-case is too specific to what we are doing.

mattmassicotte commented:
> As for the `retrieveLogin()` public accessor, I'm glad to discuss and perhaps my use-case is too specific to what we are doing

You know, I doubt it. As is, there's no way to have more than on login per authenticator, which totally makes sense. I just want to keep this PR focused so we can address this other problem well.

martindufort commented:
Yes I agree. 

We need a better construct to support multi-account authorization (multiple Login objects).

I started working on this internally in our app but internal OAuthenticator support  would be much better. 
Let's discuss the best way to lay out that capability jointhly. 

Feel free to open the previous Issue to track this.

martindufort committed:
Revert back accessor visibility

martindufort committed:
Rename GoogleDrive example to GoogleAPI
Remove unneeded `print` statements
Merge scope parameter getter into a renamed `URL+QueryParams` file
Remove unneeded commments in GoogleTests.swift

mattmassicotte commented:
Fantastic!

martindufort commented:
New release package soon ?
